File: src/main/java/galacticwars/clonewars/data/SavedDataSchemaPolicy.java
package galacticwars.clonewars.data;

import java.util.Objects;

/** Shared validation for SavedData schemas that support migration from older versions. */
public final class SavedDataSchemaPolicy {
    private SavedDataSchemaPolicy() {
    }

    public static int migrate(int persistedVersion, int currentVersion, String domain) {
        Objects.requireNonNull(domain, "domain");
        if (persistedVersion < 0) {
            throw new IllegalArgumentException("Invalid " + domain + " schema " + persistedVersion);
        }
        if (persistedVersion > currentVersion) {
            throw new IllegalArgumentException("Unsupported " + domain + " schema " + persistedVersion);
        }
        return currentVersion;
    }
}

File: src/main/kotlin/galacticwars/clonewars/data/SavedDataSchemaPolicy.kt
package galacticwars.clonewars.data

/** Shared validation for saved-data schemas that support migration from older versions. */
object SavedDataSchemaPolicy {
    @JvmStatic
    fun migrate(persistedVersion: Int, currentVersion: Int, domain: String): Int {
        require(persistedVersion >= 0) {
            "Invalid $domain schema $persistedVersion"
        }
        require(persistedVersion <= currentVersion) {
            "Unsupported $domain schema $persistedVersion"
        }
        return currentVersion
    }
}
